contrib/modelci/README.md

```diff
@@ -0,0 +1,143 @@
+        self.factors = []
+        
+    def add_record(self, r): # called when the model run, add execution details.
+        self.factors.append(r)
+        
+    def test(): # called when test
+        # can be something comparing the execution details with historical data.
+        raise NotImplementedError
+        
+    @staticmethod
+    def register(factor): # factor should be a subclass
+        assert isinstance(factor, Factor)
+        key = factor.__name__
+        assert key not in Factor.dic
+        Factor.dic[key] = factor
+    
+       def __del__(self):
+        if self.factors:
+            # write to file self.out_file
+```
+
+More factors can be integrated into the test framework, for example, a factor tracker which test the training duration can be added in the following way
+
+```python
+class TrainDurationFactor(Factor):
+    def __init__(self, threshold):
+        super(TrainDurationFactor, self).__init__('train.dura.txt')
+        self.threshold = threshold
+    
+    def test(self):
+        cur_data = _load_nparray_from_file(self.out_file)
+        his_data = _load_nparray_from_file(self.his_file)
+        diff = np.abs(cur_data - his_data) / his_data
+        if (diff > self.threshold).any():
+            raise TestError
+# register this factor tracker into the framework      
+Factor.register(TrainDurationFactor)
+```
+
+A testable model should have a file called `meta.py` with some configurations about those factors to use like
+
+```python
+# this is a demo for meta.py
+train_duration_factor = TrainDurationFactor(0.1)
+valid_duration_factor = ValidDurationFactor()
+train_memory_factor = TrainMemoryFactor()
+
+tracking_factors = [ train_duration, valid_duration_factor, train_meta_factor ]
+```
+
+Inside the model, one should call the factor trackers and add records.
+
+```python
+# configuration of some model
+import paddle
+import meta
+# ...
+
+
+for batch in batches:
+    # ...
+    duration = _get_duration_of_this_batch()
+    train_duration_factor.add_record(duration)
+    # ...
+```
+
+and the test framework will test each factor like
+
+```python
+for tracker in some_model.meta.tracking_factors:
+    some_model._run() # run and tracking_factors will collect running status
+    try:
+        tracker.test()
+    except TestError:
+        _collect_error_info
+        _ring_alarm
+```
+
+## Build a testable model
+
+The models should be placed in `./models` directory, each has a sub-directory, and a `train.xsh` script to define how to run this model. After triggering the `train.xsh`, all the data of `tracking_factors` should be created.
+
+For example, a normal model might have following logic
+
+```python
+# train.xsh
+run_train_cpu
+run_train_gpu
+```
+
+To make the testing logic stable, the testable model should assure that
+
+- fix the random seed to make result reproducible
+- just run 10 or 20 batches, and each batch takes no more than 30 seconds to limit the overall execution time
```

Comment:
I suspect 30seconds is not enough. Loss could be spiky at the beginning. I would try 10~30min and get the loss.

Reply:
I think we need eight models to cover the four application fields(image, audio, NLP, and CTR). Each model has at least two computation modes such as CPU and GPU, so the time consumption for a full test is from 3h(10min/test) to 9h on one free machine.

I wonder how long is the ideal duration for a full test.
@panyx0718